src/System.Private.ServiceModel/src/System/ServiceModel/Channels/WebSocketTransportDuplexSessionChannel.cs

```diff
@@ -536,27 +536,9 @@ private void AsyncReceiveCancelled()
                 return message;
             }
 
-            // TODO: As we're waiting blocking on a task anyway, should just call ReceiveAsync and block on that task.
             public Message Receive(TimeSpan timeout)
             {
-                bool waitingResult = _receiveTask.Task.WaitWithTimeSpan(timeout);
-                ThrowOnPendingException(ref _pendingException);
-
-                if (!waitingResult)
-                {
-                    throw FxTrace.Exception.AsError(new TimeoutException(
-                               SR.Format(SR.WaitForMessageTimedOut, timeout),
-                               TimeoutHelper.CreateEnterTimedOutException(timeout)));
-                }
-
-                Message message = GetPendingMessage();
-
-                if (message != null)
-                {
-                    StartNextReceiveAsync();
-                }
-
-                return message;
+                return ReceiveAsync(timeout).WaitForCompletionNoSpin();
```

Comment:
With WebSockets, we maintain a pending receive because it's a duplex transport channel. This means the majority of the time we won't have any data waiting for us so we don't want to spin wait. If there is data waiting, then it won't use the WaitHandle at all as the Task will complete immediately.